drivers: misc: Add support renesas_elc driver for Renesas devices

khoa-nguyen-18 opened the pull request:
The Event Link Controller (ELC) is a control circuit that assigns interrupt signals generated by built-in functions such as timers, serial interfaces, external interrupts, etc., as start triggers for other built-in functions. This allows a direct link between modules without CPU intervention.

Conceptual diagram:

<img src="https://github.com/user-attachments/assets/ebdaaba9-faff-4405-bcb3-a31f217b2ab5" width="600"/> 


The diagram below illustrates one peripheral output of the ELC. In this example, a conversion start is triggered for ADC0 Group A when the GPT0 counter overflows:

<img src="https://github.com/user-attachments/assets/01f34469-a253-49d1-a307-3ed67355a873" width="600"/> 

In this PR, we aim to:
- Define a general `renesas_elc` driver intended to support the RA, RX, and RZ series
- Add support `renesas_elc` driver for Renesas RA devices
- Add a new test suite for the Renesas ELC driver
- Add a new sample demonstrating the use of the Renesas ELC

khanhnguyenwz committed:
include: drivers: misc: Add header file for Renesas ELC driver APIs

Add header file to support Renesas ELC subsystem driver APIs

Signed-off-by: Khanh Nguyen <khanh.nguyen.wz@bp.renesas.com>

khanhnguyenwz committed:
drivers: misc: add Renesas RA ELC driver

Add support for the Renesas RA Event Link Controller, including
driver sources, Kconfig, and Devicetree bindings.

Signed-off-by: Khanh Nguyen <khanh.nguyen.wz@bp.renesas.com>
Signed-off-by: Khoa Nguyen <khoa.nguyen.xh@renesas.com>

khanhnguyenwz committed:
dts: arm: renesas: ra: add ELC node and enums for RA SoCs

Add ELC nodes to RA SoC DTS files and provide
device-specific ELC signal enums for Devicetree bindings

Signed-off-by: Khanh Nguyen <khanh.nguyen.wz@bp.renesas.com>

khanhnguyenwz committed:
tests: boards: renesas: elc: add ztests for Renesas ELC driver

Add a new test suite for the Renesas ELC driver.

Signed-off-by: Khanh Nguyen <khanh.nguyen.wz@bp.renesas.com>
Signed-off-by: Khoa Nguyen <khoa.nguyen.xh@renesas.com>

khanhnguyenwz committed:
drivers: pwm: Update Renesas RA PWM driver to integrate Renesas ELC

Update the Renesas RA PWM driver to integration
with the Renesas Event Link Controller.

Signed-off-by: Khanh Nguyen <khanh.nguyen.wz@bp.renesas.com>

khanhnguyenwz committed:
samples: boards: renesas: add Renesas ELC samples

Add a new sample demonstrating the use of the Renesas ELC

Signed-off-by: Khanh Nguyen <khanh.nguyen.wz@bp.renesas.com>
Signed-off-by: Khoa Nguyen <khoa.nguyen.xh@renesas.com>

duynguyenxa commented:
@khoa-nguyen-18 , can you add the functionality of ELC on the PR description

khoa-nguyen-18 commented:
> @khoa-nguyen-18 , can you add the functionality of ELC on the PR description

I have updated it, thank you

nordicjm commented on drivers/misc/renesas_elc/Kconfig:
```diff
@@ -0,0 +1,21 @@
+
+config RENESAS_ELC
+	bool "Renesas Event Link Controller Driver"
+	help
```
lacking depends on

khoa-nguyen-18 replied:
This is the menuconfig for the driver. We leave this config for the user to select when they need to use it. The configuration required to build the source is defined in the `Kconfig.renesas_ra_elc` file, which also includes the `depends on` condition.

nordicjm replied:
No there **needs** to be a depends on here, why is this option visible to me if I'm building for qemu x86 or an stm32? Properly guard it. Plus this is a new driver class and would need a talk/discussion/walk through at architecture review, adding label

khoa-nguyen-18 replied:
> No there needs to be a depends on here, why is this option visible to me if I'm building for qemu x86 or an stm32? Properly guard it

Sorry for the misunderstanding in my previous comment. I understand now. I'll update it before the architecture review. Thank you

nordicjm commented on samples/boards/renesas/elc/boards/ek_ra4m3.overlay:
```diff
@@ -0,0 +1,47 @@
+	status = "okay";
+};
+
+&elc{
```
missing space before bracket, here and other places, fix in whole PR

khoa-nguyen-18 replied:
thank you, I have updated it

khoa-nguyen-18 commented on drivers/misc/renesas_elc/Kconfig:
```diff
@@ -0,0 +1,21 @@
+	int "Renesas Event Link Controller initialization priority"
+	default KERNEL_INIT_PRIORITY_DEVICE
+	help
+	System initialization priority for Renesas Event Link Controller drivers.
```
Thank you, I have updated this mistake

soburi commented on drivers/misc/CMakeLists.txt:
```diff
@@ -9,5 +9,6 @@ add_subdirectory_ifdef(CONFIG_TIMEAWARE_GPIO timeaware_gpio)
 add_subdirectory_ifdef(CONFIG_DEVMUX devmux)
 add_subdirectory_ifdef(CONFIG_NORDIC_VPR_LAUNCHER nordic_vpr_launcher)
 add_subdirectory_ifdef(CONFIG_MCUX_FLEXIO mcux_flexio)
+add_subdirectory_ifdef(CONFIG_RENESAS_ELC renesas_elc)
```
If it is for RA only, it is better to add a prefix.

khoa-nguyen-18 replied:
This ELC driver is not specific to RA only; it can also be supported on RX and RZ.

soburi replied:
Ok, so you should better rename `Kconfig.renesas_ra_elc`, `renesas_ra_elc.c`, and `renesas,ra-elc.yaml`

khoa-nguyen-18 replied:
For your information, the ELC driver is a common peripheral found in many popular Renesas series such as RA, RX, and RZ. Therefore, we are defining a general ELC driver and have currently added support for RA only. We’ve created a general Kconfig for ELC, and `Kconfig.renesas_ra_elc` is included within it. For RX and RZ, we will add separate files such as `Kconfig.renesas_rx_elc`, `Kconfig.renesas_rz_elc`, and also for the source C and binding file.

soburi replied:
If the implementation can be shared, it is a good idea to make it so, but if the implementation is separate, that is, the .c files are individual, I think there is no need to define groupings. This is because only the name is common and the contents are different.

khoa-nguyen-18 replied:
The concept of the ELC driver in RA, RX, and RZ is the same, and we’ve confirmed that it works when we share the same API in `renesas_elc.h` and the handler definitions in `renesas_elc_handlers.c` across these series. However, for each series, we need to implement the .c files separately to accommodate updates from the HAL of each series.

khoa-nguyen-18 commented:
Hi @binhnguyen2434 @nhutnguyenkc, could you please take a look at this ELC support PR, as it might be related to your support plan?

khoa-nguyen-18 commented:
Add `depends on SOC_FAMILY_RENESAS_RA` for `menuconfig RENESAS_ELC`

thenguyenyf commented on drivers/misc/renesas_elc/renesas_ra_elc.c:
```diff
@@ -0,0 +1,150 @@
+static int renesas_ra_elc_software_event_generate(const struct device *dev, uint32_t event)
+{
+	struct renesas_ra_elc_data *data = (struct renesas_ra_elc_data *const)(dev)->data;
+	int err;
+
+	err = R_ELC_SoftwareEventGenerate(&data->fsp_ctrl, event);
+	if (err) {
+		return -EIO;
+	}
```
```suggestion
	fsp_err_t err;

	err = R_ELC_SoftwareEventGenerate(&data->fsp_ctrl, event);
	if (err != FSP_SUCCESS) {
		return -EIO;
	}
```

thenguyenyf replied:
Please check for other similar.

khoa-nguyen-18 replied:
Thank you. It's been updated

thenguyenyf commented on drivers/pwm/pwm_renesas_ra.c:
```diff
@@ -39,6 +40,8 @@ struct pwm_renesas_ra_data {
 	gpt_instance_ctrl_t fsp_ctrl;
 	timer_cfg_t fsp_cfg;
 	gpt_extended_cfg_t extend_cfg;
+	struct renesas_elc_dt_spec start_renesas_elc;
+	struct renesas_elc_dt_spec stop_renesas_elc;
```
It should be wrapped by RENESAS_RA_ELC
```suggestion
#ifdef CONFIG_RENESAS_RA_ELC
	struct renesas_elc_dt_spec start_renesas_elc;
	struct renesas_elc_dt_spec stop_renesas_elc;
#endif /* CONFIG_RENESAS_RA_ELC */
```

thenguyenyf commented on drivers/pwm/pwm_renesas_ra.c:
```diff
@@ -181,6 +184,15 @@ static int pwm_renesas_ra_set_cycles(const struct device *dev, uint32_t pin, uin
 		return -EIO;
 	}
 
+	/* Enables external event triggers */
+	if (data->extend_cfg.start_source != GPT_SOURCE_NONE ||
+	    data->extend_cfg.stop_source != GPT_SOURCE_NONE) {
+		err = R_GPT_Enable(&data->fsp_ctrl);
+		if (err != FSP_SUCCESS) {
+			return -EIO;
+		}
+	}
```
Should be wrapped by CONFIG_RENESAS_RA_ELC. Please recheck others similar.

khoa-nguyen-18 replied:
Agree with you. We have updated it

carlescufi commented:
This topic will be scheduled soon for the [Architecture WG meeting](https://github.com/zephyrproject-rtos/zephyr/wiki/Architecture-Working-Group).

thenguyenyf commented:
> This topic will be scheduled soon for the [Architecture WG meeting](https://github.com/zephyrproject-rtos/zephyr/wiki/Architecture-Working-Group).

Hello @carlescufi , is there any plan to schedule Arch review for this PR? It'd be nice to start reviewing for this PR soon to make it can be merged in v4.2.0.

carlescufi commented:
> > This topic will be scheduled soon for the [Architecture WG meeting](https://github.com/zephyrproject-rtos/zephyr/wiki/Architecture-Working-Group).
> 
> Hello @carlescufi , is there any plan to schedule Arch review for this PR? It'd be nice to start reviewing for this PR soon to make it can be merged in v4.2.0.

Sorry for the delay @thenguyenyf. Would next Tuesday, 27th May work for you?

thenguyenyf commented:
> > > This topic will be scheduled soon for the [Architecture WG meeting](https://github.com/zephyrproject-rtos/zephyr/wiki/Architecture-Working-Group).
> > 
> > 
> > Hello @carlescufi , is there any plan to schedule Arch review for this PR? It'd be nice to start reviewing for this PR soon to make it can be merged in v4.2.0.
> 
> Sorry for the delay @thenguyenyf. Would next Tuesday, 27th May work for you?

Yes, we can. Cc @khoa-nguyen-18 .

carlescufi commented:
> > > > This topic will be scheduled soon for the [Architecture WG meeting](https://github.com/zephyrproject-rtos/zephyr/wiki/Architecture-Working-Group).
> > > 
> > > 
> > > Hello @carlescufi , is there any plan to schedule Arch review for this PR? It'd be nice to start reviewing for this PR soon to make it can be merged in v4.2.0.
> > 
> > 
> > Sorry for the delay @thenguyenyf. Would next Tuesday, 27th May work for you?
> 
> Yes, we can. Cc @khoa-nguyen-18 .

Perfect, all set then! I will send the agenda next week.

khoa-nguyen-18 commented:
- Rebased main to resolve conflicts
- Added new `interconn` folder under `zephyr/drivers/misc/`
- Moved the driver to `zephyr/drivers/misc/interconn/renesas_elc/`

carlescufi commented:
Architecture WG meeting:

- @nashif presents some thoughts:
   1. `include/zephyr/drivers/` should strive towards having generic APIs
   2. Everything `include/zephyr/misc` should be moved outside to corresponding driver classes
   3. A new driver class can be started in `include/zephyr/drivers/` even when there's a single implementation that is vendor-specific
   4. In some cases, if it's completely hardware-specific, the code can be in the SoC folder
- @bjarki-andreasen mentions API extensions, but those are excluded from this conversation for now
- @galak suggest perhaps having a new `drivers/soc/` folder where SoC-specific drivers can live
- @teburd still sees the value in trying to abstract the differences between the different interconnect hardware peripherals, but this API is particular to the Renesas implementation

galak commented:
I think we could define a more generic API for something like hwevent across the different vendors HW that has similar functionality.

I'm guessing each of the SoCs have a similar concept in that you have some 'event' competition and that is "chained" to kick of some 'action'.  The events will be SoC specific (similar to how pinmux would be) but the chaining APIs feel like they could be common.  Feels like looking at pinmux APIs as possible abstraction of the HW specific events, but definition of common APIs would be a place to start.

carlescufi commented:
Long-term discussion: https://github.com/zephyrproject-rtos/zephyr/issues/91373

@khoa-nguyen-18 for now this PR can go in as-is.

carlescufi commented:
@thenguyenyf @nordicjm please revisit

carlescufi commented:
Apologies, I closed it by mistake.

nordicjm reviewed (approved):
had forgot I had a nack on this PR. CMake/dts files OK

KhiemNguyenT reviewed (approved):
Thanks for resolving all feedback for this new driver. LGTM.
